src/UserManager.php

```diff
@@ -0,0 +1,353 @@
+    $this->logger->info($this->formatPlural(
+      count($userIds),
+      '1 user marked for deletion.',
+      '@count users marked for deletion.'
+    ));
+  }
+
+  /**
+   * Retain users.
+   *
+   * @param array $users
+   *   The users to retain.
+   */
+  public function retainUsers(array $users) {
+    $this->validateConfig();
+    $userIdClaim = $this->moduleConfig->get('api.user_id_claim');
+    $userIdField = $this->moduleConfig->get('general.user_id_field');
+
+    $this->logger->info($this->formatPlural(
+      count($users),
+      'Retaining one user.',
+      'Retaining @count users.'
+    ));
+    $userIds = $this->getUserIds();
+    if (is_array($userIds)) {
+      foreach ($users as $user) {
+        $userId = $user[$userIdClaim] ?? NULL;
+        if (NULL === $userId) {
+          throw new \RuntimeException(sprintf('Cannot get user id (%s)', $userIdClaim));
+        }
+        try {
+          $user = $this->loadUserByProperties([$userIdField => $userId]);
+        }
+        catch (QueryException $queryException) {
+          throw new \RuntimeException(sprintf('Cannot load user by field %s', $userIdField));
+        }
+        if (NULL !== $user) {
+          $this->logger->info($this->t('Retaining user @name.', ['@name' => $user->getAccountName()]));
+          unset($userIds[$user->id()]);
+        }
+      }
+      $this->setUserIds($userIds);
+    }
+  }
+
+  /**
+   * Delete users.
+   */
+  public function deleteUsers() {
+    $method = $this->moduleConfig->get('user_cancel_method');
+    $deletedUserIds = [];
+    $userIds = $this->getUserIds();
+    foreach ($userIds as $userId) {
+      user_cancel([], $userId, $method);
+      $deletedUserIds[] = $userId;
+    }
+
+    $this->logger->info($this->formatPlural(
+      count($deletedUserIds),
+      'One user to be deleted',
+      '@count users to be deleted'
+    ));
+
+    if (!($this->options['dry-run'] ?? FALSE)) {
+      if (!empty($deletedUserIds)) {
+        $this->logger->info($this->formatPlural(
+          count($deletedUserIds),
+          'Deleting one user',
+        'Deleting @count users'
+        ));
+        $this->requestStack->getCurrentRequest()
+          // batch_process needs a route in the request (!)
+          ->attributes->set(RouteObjectInterface::ROUTE_OBJECT, new Route('<none>'));
+
+        // Process the batch created by deleteUser.
+        $batch =& batch_get();
+        $batch['progressive'] = FALSE;
+        $batch['source_url'] = 'cron';
+
+        batch_process();
+      }
+    }
+  }
+
+  /**
+   * Load user by properties.
+   *
+   * @param array $properties
+   *   The properties.
+   *
+   * @return \Drupal\user\UserInterface|null
+   *   The user if any.
+   */
+  private function loadUserByProperties(array $properties = []): ?UserInterface {
+    /** @var \Drupal\user\UserInterface[] $users */
+    $users = $this->userStorage->loadByProperties($properties);
+
+    return reset($users) ?: NULL;
```

Comment:
Would be better to bulk load users, and not do individual queries. Please see: https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Entity%21EntityStorageInterface.php/function/EntityStorageInterface%3A%3AloadByProperties/8.2.x

Reply:
Good point. I have no idea why I did it like this. I'll fix it.